New Resource: Key Vault Access Policy

sophos-jeff opened the pull request:
Hello everyone.  

This is to address #754 which allows for separating our Key Vault Policies from the Key Vault Resource.  This is also a feature I wanted as it allows me to pass in N number of policies for a Key Vault where N is not a constant.  I can now add 'count' to the policy resource to create multiple at once.

The resource looks like below allowing you to update the policy of an existing policy. 

```
resource "azurerm_key_vault_policy" "test" {
  vault_name                   = "NAME_OF_VAULT"
  vault_resource_group  = "RESOURCE_GROUP_OF_VAULT"

  object_id 		= "OBJECT_ID of service principal"
  tenant_id		= "TENANT_ID of service principal"
  application_id 	= "APPLICATION_ID if granting access to an application"

  key_permissions = [
    "create",
    "get"
  ]

  secret_permissions = [
    "get",
    "delete"
  ]
  certificate_permissions = [
    "create", 
    "delete"
  ]
}
```

This PR changes some behavior of the existing resrouce
* Minimum Access Policy blocks in a key vault resource has been reduced to 0
* Checking the access policies has been removed from the refresh of the Key Vault Resource (this is to prevent it from removing policies added by a policy resource object)


The way the new Key Vault resource interacts with these is as follows
* If the object id / tenant id pair already exists in the access policy there will be no failure and in effect that resource will be imported into the terraform state (Hence the lacking of the ability to import since it already works seamlessly) //Question. Should I add something similar to this to the doc section explaining why there isn't an import?
* If the access policy is removed outside of the terraform plan the refresh will want to add it back on a plan

I also moved the access policy definitions into its own file so there is a single place to update this.

Let me know what changes if any you guys would like me to make here.  I do apologize if my go code is sloppy, this is my first time contributing to terraform and my first time writing in go, so I am sure there will be some issues to resolve.

-Jeff

dawg89 commented:
Reviewed and tested this code. The implementation is solid, and everything worked perfectly in my real-world test. Thanks for this!

sophos-jeff commented on azurerm/resource_arm_key_vault.go:
```diff
@@ -364,50 +298,11 @@ func flattenKeyVaultSku(sku *keyvault.Sku) []interface{} {
 	return []interface{}{result}
 }
 
-func flattenKeyVaultAccessPolicies(policies *[]keyvault.AccessPolicyEntry) []interface{} {
```
This was removed due to it only being used in a refresh.  Refresh the access policies on the key vault resource causes a fight between key vault and key vault policy over ownership of the resource (basically doing this during a keyvault refresh it wants to remove resources created by a key vault policy

tiwood commented:
Any updates on this? I have some use cases for this new resource type.

Thanks for your contribution.

JunyiYi commented on azurerm/resource_arm_key_vault.go:
```diff
@@ -76,7 +76,7 @@ func resourceArmKeyVault() *schema.Resource {
 				Type:     schema.TypeList,
 				Optional: true,
-				MinItems: 1,
+				MinItems: 0,
```
Since it is optional, you might just remove this line (`MinItems: 0`).

sophos-jeff replied:
Removed.

JunyiYi commented on azurerm/resource_arm_key_vault_policy.go:
```diff
@@ -0,0 +1,221 @@
+	log.Printf("[INFO] preparing arguments for Azure ARM Policy: %s.", action)
+
+	name := d.Get("vault_name").(string)
+	//location := azureRMNormalizeLocation(d.Get("vault_location").(string))
```
If it is not used, remove this line.

sophos-jeff replied:
Removed

JunyiYi commented on azurerm/resource_arm_key_vault_policy.go:
```diff
@@ -0,0 +1,221 @@
+
+	d.SetId(*read.ID)
+
+	return resourceArmKeyVaultRead(d, meta)
```
Typically we don't read terraform resources after it is deleted, and we don't set its ID either.

sophos-jeff replied:
Added a check that we aren't deleting it.

JunyiYi commented on azurerm/resource_arm_key_vault_policy.go:
```diff
@@ -0,0 +1,222 @@
+	}
+
+	if action != keyvault.Remove {
+		d.SetId(*read.ID)
```
```go
if action != keyvault.Remove {
  d.SetId(*read.ID)
  return resourceArmKeyVaultRead(d, meta)
} else {
  return nil
}
```

sophos-jeff replied:
Good point, fixed

achandmsft commented:
@tombuildsstuff could you please merge?

tombuildsstuff commented on azurerm/resource_arm_key_vault_policy_test.go:
```diff
@@ -0,0 +1,355 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func TestAccAzureRMKeyVaultPolicy_basic(t *testing.T) {
```
can we also add an example with multiple access policies?

tombuildsstuff commented on website/docs/r/key_vault_policy.html.markdown:
```diff
@@ -0,0 +1,96 @@
+
+resource "azurerm_key_vault" "test" {
+  name                = "testvault"
+  location            = "West US"
```
_minor_ in newer examples we're making this:

```
location = "${azurerm_resource_group.test.location}"
```

so that the user only has to update this in a single location - can we update this? :)

sophos-jeff commented on azurerm/provider.go:
```diff
@@ -156,6 +156,7 @@ func Provider() terraform.ResourceProvider {
 			"azurerm_key_vault_certificate":               resourceArmKeyVaultCertificate(),
 			"azurerm_key_vault_key":                       resourceArmKeyVaultKey(),
 			"azurerm_key_vault_secret":                    resourceArmKeyVaultSecret(),
+			"azurerm_key_vault_policy":                    resourceArmKeyVaultPolicy(),
```
I actually was expecting this one and told my team I didn't want to build support for this resource into anything until the review since I thought we would need to rename it.  I really should have named this azurerm_key_vault_access_policy from the start.  I apologize for not doing this.

sophos-jeff commented:
@tombuildsstuff Thankyou for the review.  I plan on having an update tomorrow.

Out of curiosity as I am working on it do you guys generally have a policy around if you want contributors to rebase or squash commits before merging or do you simple take care of that with github at the merge itself?

Thanks again!

-Jeff

sophos-jeff commented on azurerm/resource_arm_key_vault_policy_test.go:
```diff
@@ -0,0 +1,355 @@
+  tags {
+    environment = "Production"
+	policy_object_id = "${data.azurerm_client_config.current.service_principal_object_id}"
+  }
```
So I am in compliance.
Is 2 or 4 spaces preferred?

Also it seems as if the go code is using tabs elsewhere.  Is the intention tabs for code and spaces from the multiline strings (like tests) ?

bcornils commented:
Hi @sophos-jeff I see this issue has a change request and the merge is blocked. Are you able to address this? Or do you need additional information to address the change request? Please let me know if there is anything I can do to help you.

sophos-jeff commented:
Thank you @bcornils 

I have a few questions above but working through the comments.  Already addressed about half of them locally and hope to push a new commit with all of the fixed by end of day tomorrow.  Seems like there is a lot of comments but working through them it honestly isn't to bad.  If I need more clarity on a comment from @tombuildsstuff I will comment asking for clarity.

-Jeff

sophos-jeff commented:
@tombuildsstuff @bcornils 
(Or anyone)

I have been hitting a race condition updating the access policy at the same time I update the key vault in a test. Been tearing me hair out figuring out why.

I looked into it more and found that the route-table / route resources do already have the SAME race condition. I would rather not introduce that same race condition but would like a few minutes with someone from hashicorp to go over how to avoid it as I can't figure it out right now. I would have spent much longer on it but once I realized this issue already exists I thought I would get your input. If someone could message me and we can chat about it that would be great!

I 'could' just right my tests in a way that avoids this race condition but would greatly appreciate some advice here!

pixelicous commented:
Really hope to see this coming soon, today we call the keyvault twice, one for creation the other for permission, to overcome some limitations..
However we had a state file that got diattached from the created resources, importing them now dimmed very difficult.. it tries to modify both keyvault ersources that point to the same keyvault 🤦‍♂️

pixelicous commented:
still waiting for implementation, we need this badly!!

paultyng commented:
@sophos-jeff I see the acc test you wrote up for the similar routes race, let me dig in a little and see if I can figure out whats going on.

paultyng commented on azurerm/resource_arm_route_test.go:
```diff
@@ -79,6 +79,34 @@ func TestAccAzureRMRoute_multipleRoutes(t *testing.T) {
+			{
+				Config: postConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMRouteExists("azurerm_route.test1"),
```
I believe this is failing since this route isn't in the `postConfig`?

sophos-jeff replied:
@paultyng Sorry I never actually fixed this test to fail in the proper way.  I had accidentally added it to this PR.  Let me update the test to cause the race condition I am describing properly and I will push a change shortly

paultyng replied:
Thanks, should the KeyVault tests fail as well as written? Running them now to see the race issue.

paultyng replied:
The seem to have passed for me.

sophos-jeff replied:
I believe I currently have it in a working state...but it is very very easy to break my test again. 

in the 'testAccAzureRMKeyVaultAccessPolicy_basic' function remove the following line from that tags block
```
    policy_object_id = "${data.azurerm_client_config.current.service_principal_object_id}"
```

This should break it in the same way I was experiencing before

sophos-jeff replied:
Btw I just updated the PR to cause this test to fail due to this race condition. Should be failing now.

sophos-jeff commented:
This should be ready to review again.  I believe I have addressed all the comments and made many improvements to the resource as well.  Please let me know if there is anything else you need changed or updated

katbyte commented on azurerm/resource_arm_key_vault_access_policy_test.go:
```diff
@@ -0,0 +1,397 @@
+	})
+}
+
+func testCheckAzureRMKeyVaultAccessPolicyExists(name string, policyObjectTag string, policyApplicationTag string) resource.TestCheckFunc {
```
Should `policyApplicationTag` be `policyApplicationId`

sophos-jeff replied:
It isn't the "Id" though the the index of the id in the resource object.  This was a suggestion from @tombuildsstuff to pass this through here.  I will rename it if requested.

sophos-jeff commented on azurerm/resource_arm_key_vault.go:
```diff
@@ -8,11 +8,12 @@ import (
 	"github.com/hashicorp/terraform/helper/schema"
 	"github.com/hashicorp/terraform/helper/validation"
-	"github.com/satori/go.uuid"
+	uuid "github.com/satori/go.uuid"
```
Can you provide me an example of this? It was already using the satori one here before I started working on it.  All I did here was alias it to uuid.

sophos-jeff commented on azurerm/key_vault_access_policy.go:
```diff
@@ -0,0 +1,149 @@
+package azurerm
```
Sure.  Only issue is that ignoreCaseDiffSuppressFunc is coming from the azurerm package within providers.go.  How do I import that function in or should I just copy it?

sophos-jeff commented on azurerm/resource_arm_key_vault.go:
```diff
@@ -8,11 +8,12 @@ import (
 	"time"
 
 	"github.com/Azure/azure-sdk-for-go/services/keyvault/mgmt/2016-10-01/keyvault"
+	//"github.com/davecgh/go-spew/spew"
```
Done

sophos-jeff commented on azurerm/resource_arm_key_vault.go:
```diff
@@ -214,14 +152,19 @@ func resourceArmKeyVaultCreate(d *schema.ResourceData, meta interface{}) error {
 		Tags: expandTags(tags),
 	}
 
+	//return fmt.Errorf("%s", spew.Sdump(parameters))
```
Done

sophos-jeff commented on azurerm/resource_arm_key_vault.go:
```diff
@@ -232,19 +175,9 @@ func resourceArmKeyVaultCreate(d *schema.ResourceData, meta interface{}) error {
 	if d.IsNewResource() {
 		if props := read.Properties; props != nil {
 			if vault := props.VaultURI; vault != nil {
-				log.Printf("[DEBUG] Waiting for Key Vault %q (Resource Group %q) to become available", name, resGroup)
```
Done

sophos-jeff commented on azurerm/resource_arm_key_vault.go:
```diff
@@ -232,19 +175,9 @@ func resourceArmKeyVaultCreate(d *schema.ResourceData, meta interface{}) error {
 		if props := read.Properties; props != nil {
 			if vault := props.VaultURI; vault != nil {
-				log.Printf("[DEBUG] Waiting for Key Vault %q (Resource Group %q) to become available", name, resGroup)
-				stateConf := &resource.StateChangeConf{
```
I really don't recall changing this at all.  I think this was pulled in from master at some point.  Because I don't know why I would have changed this at all.  I will change this to be back in line with master

sophos-jeff commented on azurerm/resource_arm_key_vault_access_policy.go:
```diff
@@ -0,0 +1,224 @@
+
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/mgmt/2016-10-01/keyvault"
+	"github.com/hashicorp/terraform/helper/schema"
+	uuid "github.com/satori/go.uuid"
```
Same as my other comment.  This was the package that @tombuildsstuff was using.  I would gladly use the hashicorp one though if that is what is needed

sophos-jeff commented on azurerm/resource_arm_key_vault_access_policy.go:
```diff
@@ -0,0 +1,224 @@
+	resGroup := d.Get("resource_group_name").(string)
+
+	accessPolicy := expandKeyVaultAccessPolicy(d)
+	accessPolicies := []keyvault.AccessPolicyEntry{accessPolicy}
```
Yes I will do this

sophos-jeff commented on azurerm/resource_arm_key_vault_access_policy.go:
```diff
@@ -0,0 +1,224 @@
+		return fmt.Errorf("Error making Read request on Azure KeyVault %s: %+v", name, err)
+	}
+
+	flattenedPolicy := flattenKeyVaultAccessPolicies(resp.Properties.AccessPolicies)
```
Agreed. Will update

sophos-jeff commented on azurerm/resource_arm_key_vault_access_policy_test.go:
```diff
@@ -0,0 +1,397 @@
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+	"log"
```
Yes

sophos-jeff commented on azurerm/resource_arm_key_vault_access_policy_test.go:
```diff
@@ -0,0 +1,397 @@
+		if !ok {
+			return fmt.Errorf("Not found: %s", name)
+		}
+		log.Printf("%s", spew.Sdump(rs))
```
Yes.  Thought I caught them all.  I will remove it.

sophos-jeff commented on website/docs/r/key_vault_access_policy.html.markdown:
````diff
@@ -0,0 +1,110 @@
+terraform import azurerm_key_vault_access_policy.testPolicy /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/mygroup1/providers/Microsoft.KeyVault/vaults/test-vault/objectId/00000000-0000-0000-0000-000000000000
+```
+
+```shell
````
nope.  They are different.  One is a policy object with an application id and one without it.  IT seems clunky but it is what is unique in azure land.

sophos-jeff commented on website/docs/r/key_vault_access_policy.html.markdown:
```diff
@@ -0,0 +1,110 @@
+  vault_name           = "${azurerm_key_vault.test.name}"
+  resource_group_name  = "${azurerm_key_vault.test.resource_group_name}"
+  
+  tenant_id = "d6e396d0-5584-41dc-9fc0-268df99bc610"
```
Done

sophos-jeff commented on website/docs/r/key_vault_access_policy.html.markdown:
```diff
@@ -0,0 +1,110 @@
+* `application_id` - (Optional) The object ID of an Application in Azure Active Directory.
+
+* `certificate_permissions` - (Optional) List of certificate permissions, must be one or more from
+    the following: `create`, `delete`, `deleteissuers`, `get`, `getissuers`, `import`, `list`, `listissuers`, `managecontacts`, `manageissuers`, `purge`, `recover`, `setissuers` and `update`.
```
done

sophos-jeff commented:
@katbyte Thanks for your feedback..

Specifically if we are talking about what you suggested about changing it to look like the following

```hcl
resource "azurerm_key_vault_policy" "test" {
  vault_name                   = "NAME_OF_VAULT"
  vault_resource_group  = "RESOURCE_GROUP_OF_VAULT"

access_policy = {
  object_id 		= "OBJECT_ID of service principal"
  tenant_id		= "TENANT_ID of service principal"
  application_id 	= "APPLICATION_ID if granting access to an application"

  key_permissions = [
  ]

  secret_permissions = [
  ]
  certificate_permissions = [
  ]
}
}
```

I am not sure I agree. I was looking at how it was done with route_table and route in terraform and followed that pattern for consistency across existing resources.

https://www.terraform.io/docs/providers/azurerm/r/route_table.html
https://www.terraform.io/docs/providers/azurerm/r/route.html

So unless there is a strong desire to change this I would recommend this path forward to keep consistency with existing resources.

Thanks!

I have addressed all your comments.  I am going to rerun the acceptance tests and write a few actual unit tests and I will push my changes by EOD today hopefully.

sophos-jeff commented:
This should be fully ready to review again.  I have addressed I think everything that Tom and Kat has mentioned (Either in a comment or in code) Please let me know if there is anything else I need to do to get this feature added.

goraxe commented:
just a comment so this branch appears in my github event feed

tiwood commented:
I'm really desperate to get this functionality, is there a way to help or do we just need to wait for the reviews?

sophos-jeff commented:
I think at this point it is waiting for reviews.  It is not a small PR so it would take a little bit longer...hopefully tom and kat can get to it soon, but they are both very busy and have a lot to review....best think you can do is upvote this PR in my opinion.

day-jeff commented:
@tiwood: Another option is to clone and build Jeff's branch: sophos-jeff:feature/key_vault_policy

tombuildsstuff commented:
@sophos-jeff apologies for the delayed re-review here; I'm still trying to catch up from a few weeks AFK - I'll be taking another look through this shortly :)

tombuildsstuff commented:
Key Vault Access Policy Tests pass:

```
$ acctests azurerm TestAccAzureRMKeyVaultAccessPolicy_

=== RUN   TestAccAzureRMKeyVaultAccessPolicy_basic
--- PASS: TestAccAzureRMKeyVaultAccessPolicy_basic (222.29s)
=== RUN   TestAccAzureRMKeyVaultAccessPolicy_multiple
--- PASS: TestAccAzureRMKeyVaultAccessPolicy_multiple (228.28s)
=== RUN   TestAccAzureRMKeyVaultAccessPolicy_update
--- PASS: TestAccAzureRMKeyVaultAccessPolicy_update (238.59s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	689.209s
```

```
$ acctests azurerm TestAccAzureRMKeyVaultAccessPolicy_import

=== RUN   TestAccAzureRMKeyVaultAccessPolicy_importBasic
--- PASS: TestAccAzureRMKeyVaultAccessPolicy_importBasic (225.81s)
=== RUN   TestAccAzureRMKeyVaultAccessPolicy_importMultiple
--- PASS: TestAccAzureRMKeyVaultAccessPolicy_importMultiple (218.71s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	444.557s
```

tombuildsstuff commented:
Ignoring an existing failing test in Key Vault - the tests pass:

<img width="420" alt="screenshot 2018-07-09 at 09 57 54" src="https://user-images.githubusercontent.com/666005/42437836-8f33517a-835e-11e8-91ae-1bfb86d82996.png">

sophos-jeff commented:
Not at all Tom.  Thanks for getting this merged :D Glad to see this feature available soon :D

jmeisner3 commented:
@sophos-jeff thanks a lot for working on this!

pixelicous commented:
@tombuildsstuff amazing work! finally this is merged!! 😋  👍 🥇

pixelicous commented:
@tombuildsstuff Great work! This would be very useful to us and anyone else using azure, finally this is merged!! 😋  👍 🥇
